### PR TITLE
TRPO, PPO1 - fix EpThisIter, EpisodesSoFar and TimestepsSoFar

### LINKED ISSUE #164: [question] How is TimestepsSoFar computed?
Hi,

My stable-baselines version is 2.3. I'm trying to understand how TimestepsSoFar is computed for TRPO. In the logs after I start a training, first stats are as following:

```
----------------------------------
| EpLenMean       | 128          |
| EpRewMean       | -840         |
| EpThisIter      | 1            |
| EpisodesSoFar   | 1            |
| TimeElapsed     | 18.3         |
| TimestepsSoFar  | 128          |
| entloss         | 0.04307969   |
| entropy         | 4.307969     |
| ev_tdlam_before | -0.00229     |
| meankl          | 0.0095857065 |
| optimgain       | 0.1366807    |
| surrgain        | 0.09360102   |
----------------------------------
```

Then, second one:

```
---------------------------------
| EpLenMean       | 192         |
| EpRewMean       | -1.24e+03   |
| EpThisIter      | 1           |
| EpisodesSoFar   | 2           |
| TimeElapsed     | 23.9        |
| TimestepsSoFar  | 384         |
| entloss         | 0.04303209  |
| entropy         | 4.3032093   |
| ev_tdlam_before | 1.61e-06    |
| meankl          | 0.010002808 |
| optimgain       | 0.14815919  |
| surrgain        | 0.105127096 |
---------------------------------
``` 

3rd trace: 

```
---------------------------------
| EpLenMean       | 256         |
| EpRewMean       | -1.67e+03   |
| EpThisIter      | 1           |
| EpisodesSoFar   | 3           |
| TimeElapsed     | 29.6        |
| TimestepsSoFar  | 768         |
| entloss         | 0.042882647 |
| entropy         | 4.2882648   |
| ev_tdlam_before | 0           |
| meankl          | 0.010393932 |
| optimgain       | 0.16137204  |
| surrgain        | 0.118489385 |
---------------------------------
```

You can see that `TimestepsSoFar` seems to double every time although batch size is 128. The problem I have happens with a custom env (tried with CartPole, it seems to increment correctly). 

What could be the reason for that?

### PR DESCRIPTION
**Pull request by antoine-galataud**
fix for #164 which is about TimestepsSoFar computation in TRPO, but I also took the opportunity to fix EpThisIter and EpisodesSoFar that were not reflecting reality. I think this went under radar because episode length is lesser than batch size for some environments. 

Also, I applied the same fix for PPO1. Didn't check for other algos. 

Tested on several classic environments, custom environment with episode length >> batch size, with and without MPI.

Fixes #164

### CODE REVIEW

**Comment by araffin** [comment]
Hello,
Thanks for your PR =)
I do not have the time to have a deeper look into that issue for now, will try before next week.
Quick question: I see that you fixed the algorithms, wouldn't that be easier to fix the `traj_segment_generator` instead? (that both algorithms use I think)

As described in [CONTRIBUTING.md](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) please do not forget to update the changelog and add your name in the contributors list at the bottom ;)

**Comment by antoine-galataud** [comment]
@araffin I updated the changelog. Thanks!

**Comment by antoine-galataud** [comment]
Problem fixed. The issue was with PPO1 only, and was affecting `steps` calculation, doubled when episode was done. Just moved global counter update after `steps` calculation

**Comment by antoine-galataud** [comment]
The problem was also there for TRPO. I fixed it and tested it, works fine now.

**Comment by araffin** [comment]
I took a deeper look into the problem, and I really think we should rather fix `traj_segment_generator` by adding `cur_ep_len = 0` on line 78 (+ a comment of why we do that).

The solution you propose works but it is not super clear and makes duplicated code (between PPO1 and TRPO). Also, fixing  `traj_segment_generator` apparently requires only one line.

**Comment by antoine-galataud** [comment]
@araffin it does fix calculation of `TimestepsSoFar`. However `EpThisIter` and `EpisodesSoFar` are still wrong (at least when episode length > batch size, didn't test in other case).

**Comment by antoine-galataud** [comment]
@araffin I'm changing my fix by taking so to set  `cur_ep_len = 0` as suggested. It indeed makes the code clearer.

**Comment by antoine-galataud** [comment]
@araffin fix changed as discussed. I leave EpThisIter and EpisodesSoFar aside.

**Comment by araffin** [comment]
>I leave EpThisIter and EpisodesSoFar aside.

I'm currently thinking about how to solve it the cleanest way ;)
Btw, you introduced a small regression: policy_kwargs are not passed anymore to the policy class.

**Comment by araffin** [comment]
Ok, it seems that removing lines 65 to 67 of `utils.py` should solve the issue, no? 
(I'm currently checking were those variable are used, apparently only for logging)

**Comment by antoine-galataud** [comment]
@araffin I was trying with removing `ep_lens = [cur_ep_len]` on line 67 and it seems to work. But then you get the warning that line was made for preventing (mean of empty slice). 
Also, probably low prio, but variable names in this file are a bit confusing. Like `cur_ep_len` should be renamed `cur_it_len` now (it = iteration), don't you think? Also `total_timesteps` could be `ep_timesteps`.

**Comment by araffin** [comment]
>But then you get the warning that line was made for preventing (mean of empty slice).

Yes, I had the same thought. But it can if solved using a simple `if len(lenbuffer) > 0` before displaying the mean, so it shows stats only when it is meaningful.

>Like cur_ep_len should be renamed cur_it_len now (it = iteration), don't you think

Totally agree, maybe more `current_it_len` and `current_it_timesteps`? (because `total_timesteps` can contain several episodes).

**Comment by antoine-galataud** [comment]
@araffin I applied discussed changes.

File: stable_baselines/ppo1/pposgd_simple.py [hunk 1]
```diff
@@ -297,8 +297,10 @@ def learn(self, total_timesteps, callback=None, seed=None, log_interval=100, tb_
                     lens, rews = map(flatten_lists, zip(*listoflrpairs))
                     lenbuffer.extend(lens)
                     rewbuffer.extend(rews)
-                    logger.record_tabular("EpLenMean", np.mean(lenbuffer))
-                    logger.record_tabular("EpRewMean", np.mean(rewbuffer))
+                    if len(lenbuffer) > 0:
```
Comment (araffin) [inline stable_baselines/ppo1/pposgd_simple.py]: Minor: I was wondering, when `len(lenbuffer) > 0`, len(rewbuffer) must be > 0 no?
If so, we can simplify the check ;)

Comment (antoine-galataud) [reply stable_baselines/ppo1/pposgd_simple.py]: True, it was more a matter of long term code stability. Will do the change.

**Comment by antoine-galataud** [comment]
Thank you @araffin

**Pull request merged**